src/middleware/tests/auth.middleware.test.ts

```diff
@@ -0,0 +1,149 @@
+        AuthService.validateJwt = jest.fn().mockImplementation(<AccessToken>() => null)
+        await isAuthorized(req, res, next)
+
+        expect(res.status).toHaveBeenCalledWith(401)
+        expect(next).toHaveBeenCalledTimes(0)
+      })
+    })
+  })
+
+  describe('isValidRefresToken', () => {
+    describe('Validate via cookies', () => {
+      beforeEach(() => (req.headers.authorization = undefined)) // no header on cookies requests
+      describe('Authorization Successful', () => {
+        beforeEach(async () => {
+          AuthService.validateJwt = jest.fn().mockImplementation(<RefreshToken>() => expectedRefreshToken)
+          await isValidRefreshToken(req, res, next)
+        })
+
+        test('Refresh token added to req', () => expect(req.refreshUser).toEqual(expectedRefreshToken))
+        test('Next called with no parameters', () => expect(next).toBeCalledWith())
+      })
+
+      describe('Authorization Failure', () => {
+        beforeEach(async () => {
+          AuthService.validateJwt = jest.fn().mockImplementation(<RefreshToken>() => null)
+          await isValidRefreshToken(req, res, next)
+        })
+
+        test('Status set to 401', () => expect(res.status).toHaveBeenCalledWith(401))
+        test('Unauthorized returned as response', () => expect(res.json).toHaveBeenCalledWith(Unauthorized))
+        test('Next is not called', () => expect(next).toHaveBeenCalledTimes(0))
+      })
+    })
+
+    describe('Validate via header', () => {
+      describe('Authorization Successful', () => {
+        beforeEach(async () => {
+          AuthService.validateJwt = jest.fn().mockImplementation(<RefreshToken>() => expectedRefreshToken)
+          await isValidRefreshToken(req, res, next)
+        })
+
+        test('Refresh token added to req', () => expect(req.refreshUser).toEqual(expectedRefreshToken))
+        test('Next called with no parameters', () => expect(next).toBeCalledWith())
+      })
+
+      describe('Authorization Failure', () => {
+        beforeEach(async () => {
+          AuthService.validateJwt = jest.fn().mockImplementation(<RefreshToken>() => null)
+          await isValidRefreshToken(req, res, next)
+        })
+
+        test('Status set to 401', () => expect(res.status).toHaveBeenCalledWith(401))
+        test('Unauthorized returned as response', () => expect(res.json).toHaveBeenCalledWith(Unauthorized))
+        test('Next is not called', () => expect(next).toHaveBeenCalledTimes(0))
+      })
+    })
+  })
+
+  describe('isRefreshNearingExpiration', () => {
+    beforeEach(() => {
+      environment.refreshTokenExpirationBufferSeconds = 20 // seconds
+    })
+    test('Expiration within buffer acts as expired', async () => {
+      const nowInEpoch = Math.round(Date.now() / 1000)
+      req.refreshUser = { exp: nowInEpoch + 10 } // expires 10 seconds from now
+
+      await isRefreshNearingExpiration(req, res, next)
+
+      expect(res.status).toBeCalledWith(401)
```

Comment:
Still not sure if throwing a 401 here is the right move - they are still authorized at this point. Letting the client know somehow and handling it there likely is a better approach

Reply:
Yeah this is where we briefly talked about letting the client know that the token is expiring without necessarily causing the token to be treated as expired prematurely. Maybe this can be done in the form of a custom response header that the client checks for?